### PR TITLE
#2719 Unification explorer + collections into one

### LINKED ISSUE #2719: Unification explorer + collections into one
Based on the discussion from #2511 

For the sake and easier maintenance would be superior nice to have merged explorer and collections, so we don't need to maintain these two components as extra. It would simplify architecture in @petersopko words, which I can agree, keep it simple stupid.

The flow would look like when a user searches for "something" it can drive it to tabs, where the first highlighted tab could be collectables, other items and once we will have #2112  we can add that one as well    

>  we could merge the Gallery + Collections under one tab in the navbar (Explore) and have it separated by tabs within the search section

image

> As still, we haven't thought what merger of explore+collections feed would look like I'm open to new ideas and sense this is the time

### PR DESCRIPTION
**Pull request by prachi00**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

### What's new?

- [x] PR closes #2719
- [ ] 
1st version, let me know any additions or bugs, I am still testing it, feel free to let me know your suggestions about UI/UX. 

### Before submitting Pull Request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [ ] I've didn't break any original functionality
- [x] I've posted a screenshot of demonstrated change in this PR

### Optional

- [ ] I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
- [ ] I've tested PR on mobile and everything seems works
- [ ] I found edge cases
- [ ] I've written some unit tests 🧪

### Had issue bounty label?

- [x] Fill up your KSM address: 
[Payout](https://beta.kodadot.xyz/transfer/?target=EzGc4s9PgCPx1YnF3fqzhLzVHpHMTL4LWPScwpDrR8JKgSU)

### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot
Screenshot 2022-04-09 at 4 30 17 PM

- [x] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

### CODE REVIEW

**Comment by prachi00** [comment]
> **Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz). 👇 _ Let's make a quick check before the contribution.
> 
> ### PR type
> * [ ]  Bugfix
> * [x]  Feature
> * [ ]  Refactoring
> 
> ### What's new?
> * [x]  PR closes [Unification explorer + collections into one  #2719](https://github.com/kodadot/nft-gallery/issues/2719)
> * [ ]
>   1st version, let me know any additions or bugs, I am still testing it, feel free to let me know your suggestions about UI/UX.
PS. only done for rmrk 
> 
> ### Before submitting Pull Request, please make sure:
> * [x]  My contribution builds **clean without any errors or warnings**
> * [x]  I've merged recent default branch -- **main** and I've no conflicts
> * [x]  I've tried to respect high code quality standards
> * [ ]  I've didn't break any original functionality
> * [x]  I've posted a screenshot of demonstrated change in this PR
> 
> ### Optional
> * [ ]  I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
> * [ ]  I've tested PR on mobile and everything seems works
> * [ ]  I found edge cases
> * [ ]  I've written some unit tests 🧪
> 
> ### Had issue bounty label?
> * [x]  Fill up your KSM address:
>   [Payout](https://beta.kodadot.xyz/transfer/?target=EzGc4s9PgCPx1YnF3fqzhLzVHpHMTL4LWPScwpDrR8JKgSU)
> 
> ### Community participation
> * [x]  [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)
> 
> ### Screenshot
> Screenshot 2022-04-09 at 4 30 17 PM
> 
> * [x]  My fix has changed **something** on UI; a screenshot is best to understand changes for others.

File: components/rmrk/Collection/List/CollectionList.vue [hunk 1]
```diff
@@ -1,21 +1,16 @@
...
-    </Search>
+    <b-field>
+      <Pagination
+        hasMagicBtn
+        simple
+        replace
+        preserveScroll
+        :total="total"
+        v-model="currentValue"
+        :per-page="first" />
```
Comment (vikiival) [inline components/rmrk/Collection/List/CollectionList.vue:12]:
```suggestion
        :perPage="first" />
```

File: components/rmrk/ExploreLayout/ExploreLayout.vue [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  Prop,
+  VModel,
+  Vue,
+  Watch,
+} from 'nuxt-property-decorator'
+import CollectionList from '@/components/rmrk/Collection/List/CollectionList.vue'
+import Collections from '@/components/rmrk/Gallery/Collections.vue'
+
+import PrefixMixin from '~/utils/mixins/prefixMixin'
+import shouldUpdate from '~/utils/shouldUpdate'
```
Comment (vikiival) [inline components/rmrk/ExploreLayout/ExploreLayout.vue:26]:
```suggestion
import PrefixMixin from '@/utils/mixins/prefixMixin'
import shouldUpdate from '@/utils/shouldUpdate'
```

File: components/rmrk/ExploreLayout/ExploreLayout.vue [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+const components = {
+  CollectionList,
+  Collections,
+}
+
+@Component<ExploreLayout>({
+  components,
+})
+export default class ExploreLayout extends mixins(PrefixMixin) {
+  @Prop({ type: String, default: 'default' }) type!: string
```
Comment (vikiival) [inline components/rmrk/ExploreLayout/ExploreLayout.vue:37]: Use rather `prefixMixin`

File: components/rmrk/ExploreLayout/ExploreLayout.vue [hunk 4]
```diff
@@ -0,0 +1,59 @@
+<template>
+  <div>
+    <b-tabs v-model="selectedTab" size="is-small">
+      <b-tab-item label="COLLECTION" value="COLLECTION">
+        <CollectionList v-if="type === 'rmrk'" />
+        <Collections v-if="type !== 'rmrk'" />
```
Comment (vikiival) [inline components/rmrk/ExploreLayout/ExploreLayout.vue:6]: I'll open issue for this

**Review by vikiival — CHANGES_REQUESTED** [review]

**Comment by vikiival** [comment]
Code is oki, but: 

1. Would personally prefer first gallery then collections
Screenshot 2022-04-10 at 10 42 39

2. You made some magical rewrites so I am not sure
Screenshot 2022-04-10 at 10 44 47

3. You are going to make a breaking route change so any `/rmrk/collections` `/rmrk/gallery` will become broken (sort of proxy would be more than welcomed :))

Screenshot 2022-04-10 at 10 46 26

**Comment by prachi00** [comment]
@vikiival did the changes you mentioned, for the proxy part wrote a redirect `serverMiddleware`, seems to work fine
https://nuxtjs.org/docs/configuration-glossary/configuration-servermiddleware#servermiddleware-vs-middleware

File: components/rmrk/ExploreLayout/ExploreLayout.vue (hunk 4 above)
Comment (vikiival) [reply components/rmrk/ExploreLayout/ExploreLayout.vue:7]: https://github.com/kodadot/nft-gallery/issues/2817

File: components/rmrk/ExploreLayout/ExploreLayout.vue (hunk 3 above)
Comment (vikiival) [reply components/rmrk/ExploreLayout/ExploreLayout.vue:29]: not resolved :D

File: components/rmrk/Gallery/Search/SearchBar.vue [hunk 5]
```diff
@@ -401,7 +400,7 @@ export default class SearchBar extends mixins(
   redirectToGalleryPageIfNeed() {
     if (this.$route.name === 'index') {
       this.$router.replace({
-        name: 'rmrk-gallery',
+        path: '/rmrk/explore',
```
Comment (vikiival) [inline components/rmrk/Gallery/Search/SearchBar.vue:403]: why change  ?

you can still use imo `rmrk-explore`

File: middleware/redirects.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
+export default function (req, res, next) {
+  const redirects = [
```
Comment (vikiival) [inline middleware/redirects.ts:2]: and what about others?

File: nuxt.config.js [hunk 7]
```diff
@@ -128,6 +128,8 @@ export default defineNuxtConfig({
     middleware: ['prefix'],
   },
 
+  serverMiddleware: ['~/middleware/redirects.ts'],
```
Comment (vikiival) [inline nuxt.config.js:131]: does not work

**Review by vikiival — CHANGES_REQUESTED** [review]

**Comment by vikiival** [comment]
Screenshot 2022-04-11 at 09 52 43

looks like the redirect does not work

**Comment by vikiival** [comment]
Also I am sorry for the confusion with this. 

Screenshot 2022-04-11 at 09 57 17

It should be collection first, gallery second (my bad :/)

**Comment by prachi00** [comment]
Screenshot 2022-04-11 at 1 40 19 PM
@vikiival weird, i see it as resolved

**Comment by vikiival** [comment]
> weird, i see it as resolved

But you don't use it

**Comment by prachi00** [comment]
@vikiival redirect seems to work on local, checking why it isnt working on netlify, probably need to add it in netlify config as well

**Comment by prachi00** [comment]
> middleware

@vikiival then we would have to use just `middleware` right instead of `serverMiddleware` as we would need to access prefix?

**Comment by vikiival** [comment]
Yup

That server middleware does not work on deployed app https://github.com/kodadot/nft-gallery/pull/2812#issuecomment-1094673541

**Comment by yangwao** [comment]
>> It should be collection first, gallery second (my bad :/)

Or we can put it into preference which should be shown first?

Hence we don't have working filters for collections, let's put gallery now there welp

Once we will have filters on collections and would be finished
- #1357 

We can make collections showing first.
So right now make gallery first as focus filters there:)

**Comment by prachi00** [comment]
> > > It should be collection first, gallery second (my bad :/)
> 
> Or we can put it into preference which should be shown first?
> 
> Hence we don't have working filters for collections, let's put gallery now there welp
> 
> Once we will have filters on collections and would be finished
> 
> * [Do not show empty collections - Better filters on Collections view  #1357](https://github.com/kodadot/nft-gallery/issues/1357)
> 
> We can make collections showing first. So right now make gallery first as focus filters there:)

@yangwao okay sure, can we create another issue for adding it in settings? I'll work on it :)

File: middleware/redirects.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+export default function ({ store, redirect, route }): void {
+  const prefix = store.getters.currentUrlPrefix
+
```
Comment (vikiival) [inline middleware/redirects.ts:3]: missing check for starts with prefix

File: components/rmrk/ExploreLayout/ExploreLayout.vue [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+const components = {
+  CollectionList,
+  Collections,
+}
+
+@Component<ExploreLayout>({
+  components,
+})
+export default class ExploreLayout extends Vue {
+  @Prop({ type: String, default: 'default' }) type!: string
```
Comment (vikiival) [inline components/rmrk/ExploreLayout/ExploreLayout.vue:27]:
```suggestion
export default class ExploreLayout extends mixins(PrefixMixin) {
```

File: components/rmrk/ExploreLayout/ExploreLayout.vue [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
```
Comment (vikiival) [inline components/rmrk/ExploreLayout/ExploreLayout.vue:28]:
```suggestion
get type() {
  return this.urlPrefix
}
```

File: pages/errorcollection.vue [hunk 11]
```diff
@@ -10,11 +10,7 @@
         >creating bug issue with steps to reproduce and screenshot.</a
       >
     </p>
-    <b-button
-      tag="nuxt-link"
-      to="/rmrk/collections"
-      type="is-primary"
-      class="my-5">
+    <b-button tag="nuxt-link" to="/rmrk/explore" type="is-primary" class="my-5">
```
Comment (vikiival) [inline pages/errorcollection.vue:13]:
```suggestion
    <b-button tag="nuxt-link" to=`/${type}/explore` type="is-primary" class="my-5">
```

**Review by vikiival — CHANGES_REQUESTED** [review]
Small stuff and ready to go 🚀

File: middleware/redirects.ts (hunk 8 above)
Comment (prachi00) [reply middleware/redirects.ts:3]: I dont get this one, starts with?

Comment (prachi00) [reply middleware/redirects.ts:3]: @vikiival

Comment (vikiival) [reply middleware/redirects.ts:3]: like every route has name 

`/rmrk/gallery` `rmrk-gallery`
`/bsx/gallery` `bsx-gallery`
`/spotlight` `spotlight`

so if someone adds route 
`/collections` your current impl will redirect him/her to `/${prefix}/explore`

**Review by roiLeo — COMMENTED** [review]
- component `ExploreLayout` is in `/rmrk/ExploreLayout/` wouldn't it be better to remove it from the file since it is no longer related with rmrk?
- it doesn't risk to break all the links we had before (shared links etc..) on the `/rmrk/gallery` or `/rmrk/collections` page?

**Comment by prachi00** [comment]
> * component `ExploreLayout` is in `/rmrk/ExploreLayout/` wouldn't it be better to remove it from the file since it is no longer related with rmrk?
> * it doesn't risk to break all the links we had before (shared links etc..) on the `/rmrk/gallery` or `/rmrk/collections` page?

I guess we can remove the rmrk route after the collection part is fixed with #2817 so that it is truly common? 

For the breakage part, we handled redirection through middleware so it should be fine

File: components/rmrk/Gallery/GalleryCardList.vue [hunk 12]
```diff
@@ -31,8 +34,8 @@ const components = {
 
 @Component({ components })
 export default class GalleryCardList extends Vue {
-  @Prop({ default: '/rmrk/gallery' }) public route!: string
-  @Prop({ default: 'rmrk/gallery' }) public link!: string
+  @Prop({ default: '/rmrk/explore?tab=GALLERY' }) public route!: string
+  @Prop({ default: '/rmrk/explore?tab=GALLERY' }) public link!: string
```
Comment (roiLeo) [inline components/rmrk/Gallery/GalleryCardList.vue:38]:
```suggestion
  @Prop({ default: 'rmrk/explore?tab=GALLERY' }) public link!: string
```

File: components/shared/LinkResolver.vue [hunk 13]
```diff
@@ -21,7 +21,7 @@ import isShareMode from '@/utils/isShareMode'
 
 @Component
 export default class LinkResolver extends mixins(InlineMixin) {
-  @Prop({ default: '/rmrk/gallery' }) public route!: string
+  @Prop({ default: '/rmrk/explore?tab=GALLERY' }) public route!: string
   @Prop({ default: 'rmrk/gallery' }) public link!: string
```
Comment (roiLeo) [inline components/shared/LinkResolver.vue:25]:
```suggestion
  @Prop({ default: 'rmrk/explore?tab=GALLERY' }) public link!: string
```

**Review by roiLeo — COMMENTED** [review]
typo

**Comment by vikiival** [comment]
So how is it going? @prachi00  :)
I see some unresolved stuff.

**Comment by prachi00** [comment]
> So how is it going? @prachi00 :) I see some unresolved stuff.

@vikiival forgot to mark, have resolved them

File: pages/westend/explore.vue [hunk 14]
```diff
@@ -1,22 +1,24 @@
 <template>
-  <Collections />
+  <ExploreLayout />
 </template>
 
 <script lang="ts">
 import { Component, Vue } from 'nuxt-property-decorator'
-import Collections from '@/components/rmrk/Gallery/Collections.vue'
+import ExploreLayout from '@/components/rmrk/ExploreLayout/ExploreLayout.vue'
```
Comment (vikiival) [inline pages/westend/explore.vue:7]: why not async?

**Comment by vikiival** [comment]
pay 200 usd

**Comment by vikiival** [comment]
Looks like since this PR the main is broken

**Comment by prachi00** [comment]
it is? was working fine for me though

**Comment by vikiival** [comment]
last working state I have is 39fded43252464fdab42a0c75b854c779b650ad0

**Pull request merged**